docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
```

Comment:
Why do you need this? There's no mechanism for tying changes in the history store (e.g. new frecent sites after a sync) to this listener.

I don't think we need to live-update top sites UI as things change. Top Sites won't really change while you're looking at them (the only reason it can is if sync runs in the background and significantly changes the frecent scores).

Reply:
TopSiteFeature will be a view bound feature is it needs to inform the UI when to display content. We have use cases currently in Fenix where you can use other UI menus to alter your top sites while you can see the top sites (see screenshot below).

For this reason we need to notify the storage observers to update the top sites list that can be displayed on the visible UI if there is one.

This is also when we added the concept of a `TopSitesUseCase` to act on the storage and not on the view-bound feature.

<img width="356" alt="Screen Shot 2020-07-30 at 7 23 54 PM" src="https://user-images.githubusercontent.com/1370580/88983898-3a1d0b80-d29a-11ea-95f2-cd4eb81d398a.png">

Reply:
I see, so the feature/use-case/whatever will trigger this observer method whenever something is removed/added via the app? That seems fine.

The way I first read that, it felt like this will triggered on _any_ storage changes, which I don't think you have visibility into.

Reply:
> I see, so the feature/use-case/whatever will trigger this observer method whenever something is removed/added via the app? That seems fine.

Correct! This is an existing pattern we have across our components that touch the UI. For example, [`FindInPageView`](https://github.com/mozilla-mobile/android-components/blob/master/components/feature/findinpage/src/main/java/mozilla/components/feature/findinpage/view/FindInPageView.kt).